<a href="https://colab.research.google.com/github/MrJenish/woc5.0-ml-weather-prediction-case-study-Jenish-Mangukiya/blob/main/WOC5_0_ML_Checkpoint_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import export_text
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [2]:
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/weather.csv")

In [3]:
df.shape

(1461, 5)

In [4]:
df.weather.value_counts()

rain       641
sun        640
fog        101
drizzle     53
snow        26
Name: weather, dtype: int64

In [5]:
df.weather.replace(['drizzle','rain','sun','snow','fog'],[0,1,2,3,4],inplace=True)

In [6]:
df.weather.value_counts()

1    641
2    640
4    101
0     53
3     26
Name: weather, dtype: int64

### **Model : 1** 

In [7]:
input1 = df.drop(['weather'],axis=1)
output1 = df.weather
x_train1, x_test1, y_train1, y_test1 = train_test_split(input1,output1,test_size=0.2,random_state=42)

In [8]:
print('x_train1 size : ',str(x_train1.shape))
print('x_test1 size : ',str(x_test1.shape))
print('y_train1 size : ',str(y_train1.shape))
print('y_test1 size : ',str(y_test1.shape))

x_train1 size :  (1168, 4)
x_test1 size :  (293, 4)
y_train1 size :  (1168,)
y_test1 size :  (293,)


In [9]:
model1 = tree.DecisionTreeClassifier()
model1.fit(x_train1,y_train1)
print("Accuracy of 1st model : {:.2f}%".format(model1.score(x_test1,y_test1)*100))

Accuracy of 1st model : 73.04%


### **Model : 2**

In [10]:
x_train2, x_test2, y_train2, y_test2 = train_test_split(input1,output1,test_size=0.2,random_state=42)
print('x_train2 size : ',str(x_train2.shape))
print('x_test2 size : ',str(x_test2.shape))
print('y_train2 size : ',str(y_train2.shape))
print('y_test2 size : ',str(y_test2.shape))

x_train2 size :  (1168, 4)
x_test2 size :  (293, 4)
y_train2 size :  (1168,)
y_test2 size :  (293,)


In [11]:
model2 = tree.DecisionTreeClassifier(min_samples_leaf=1)
model2.fit(x_train2,y_train2)
print("Accuracy of 2nd model : {:.2f}%".format(model2.score(x_test2,y_test2)*100))

Accuracy of 2nd model : 75.09%


### **Model : 3**

In [12]:
x_train3, x_test3, y_train3, y_test3 = train_test_split(input1,output1,test_size=0.2,random_state=42)
print('x_train3 size : ',str(x_train3.shape))
print('x_test3 size : ',str(x_test3.shape))
print('y_train3 size : ',str(y_train3.shape))
print('y_test3 size : ',str(y_test3.shape))

x_train3 size :  (1168, 4)
x_test3 size :  (293, 4)
y_train3 size :  (1168,)
y_test3 size :  (293,)


In [13]:
model3 = tree.DecisionTreeClassifier(min_samples_leaf=5)
model3.fit(x_train3,y_train3)
print("Accuracy of 3rd model : {:.2f}%".format(model3.score(x_test3,y_test3)*100))

Accuracy of 3rd model : 78.84%


### **Model : 4**

In [14]:
x_train4, x_test4, y_train4, y_test4 = train_test_split(input1,output1,test_size=0.2,random_state=42)
print('x_train4 size : ',str(x_train4.shape))
print('x_test4 size : ',str(x_test4.shape))
print('y_train4 size : ',str(y_train4.shape))
print('y_test4 size : ',str(y_test4.shape))

x_train4 size :  (1168, 4)
x_test4 size :  (293, 4)
y_train4 size :  (1168,)
y_test4 size :  (293,)


In [15]:
model4 = tree.DecisionTreeClassifier(max_depth=3)
model4.fit(x_train4,y_train4)
print("Accuracy of 4th model : {:.2f}%".format(model4.score(x_test4,y_test4)*100))

Accuracy of 4th model : 83.28%


### **Model : 5**

In [16]:
x_train5, x_test5, y_train5, y_test5 = train_test_split(input1,output1,test_size=0.2,random_state=42)
print('x_train5 size : ',str(x_train5.shape))
print('x_test5 size : ',str(x_test5.shape))
print('y_train5 size : ',str(y_train5.shape))
print('y_test5 size : ',str(y_test5.shape))

x_train5 size :  (1168, 4)
x_test5 size :  (293, 4)
y_train5 size :  (1168,)
y_test5 size :  (293,)


In [17]:
model5 = tree.DecisionTreeClassifier(max_depth=3,min_samples_leaf=1)
model5.fit(x_train5,y_train5)
print("Accuracy of 5th model : {:.2f}%".format(model5.score(x_test5,y_test5)*100))

Accuracy of 5th model : 83.28%


### **Model : 6**

In [18]:
x_train6, x_test6, y_train6, y_test6 = train_test_split(input1,output1,test_size=0.2,random_state=42)
sc = StandardScaler()
x_train6 = sc.fit_transform(x_train6)
x_test6 = sc.transform(x_test6)
print('x_train6 size : ',str(x_train6.shape))
print('x_test6 size : ',str(x_test6.shape))
print('y_train6 size : ',str(y_train6.shape))
print('y_test6 size : ',str(y_test6.shape))

x_train6 size :  (1168, 4)
x_test6 size :  (293, 4)
y_train6 size :  (1168,)
y_test6 size :  (293,)


In [19]:
model6 = tree.DecisionTreeClassifier(min_samples_leaf=5)
model6.fit(x_train6,y_train6)
print("Accuracy of 6th model : {:.2f}%".format(model6.score(x_test6,y_test6)*100))

Accuracy of 6th model : 78.84%


### **Model : 7**

In [20]:
x_train7, x_test7, y_train7, y_test7 = train_test_split(input1,output1,test_size=0.2,random_state=42)
sc = StandardScaler()
x_train7 = sc.fit_transform(x_train7)
x_test7 = sc.transform(x_test7)
print('x_train7 size : ',str(x_train7.shape))
print('x_test7 size : ',str(x_test7.shape))
print('y_train7 size : ',str(y_train7.shape))
print('y_test7 size : ',str(y_test7.shape))

x_train7 size :  (1168, 4)
x_test7 size :  (293, 4)
y_train7 size :  (1168,)
y_test7 size :  (293,)


In [21]:
model7 = tree.DecisionTreeClassifier(max_depth = 3,min_samples_leaf=5)
model7.fit(x_train7,y_train7)
print("Accuracy of 7th model : {:.2f}%".format(model7.score(x_test7,y_test7)*100))

Accuracy of 7th model : 83.28%


### **Model : 8**

In [22]:
Q1=df.quantile(0.25)
Q3=df.quantile(0.75)
IQR=Q3-Q1
df=df[~((df<(Q1-1.5*IQR))|(df>(Q3+1.5*IQR))).any(axis=1)]

In [23]:
x_train8, x_test8, y_train8, y_test8 = train_test_split(input1,output1,test_size=0.2,random_state=42)
x_train8 = sc.fit_transform(x_train8)
x_test8 = sc.transform(x_test8)
print('x_train8 size : ',str(x_train8.shape))
print('x_test8 size : ',str(x_test8.shape))
print('y_train8 size : ',str(y_train8.shape))
print('y_test8 size : ',str(y_test8.shape))

x_train8 size :  (1168, 4)
x_test8 size :  (293, 4)
y_train8 size :  (1168,)
y_test8 size :  (293,)


In [24]:
model8 = tree.DecisionTreeClassifier(max_depth=3)
model8.fit(x_train8,y_train8)
print("Accuracy of 8th model : {:.2f}%".format(model8.score(x_test8,y_test8)*100))

Accuracy of 8th model : 83.28%


In [25]:
def predictWeather(x):
  predictedWeather = model8.predict(x)
  if predictedWeather==0:
    return "Drizzle"
  if predictedWeather==1:
    return "Rain"
  if predictedWeather==2:
    return "Sun"
  if predictedWeather==3:
    return "Snow"
  if predictedWeather==4:
    return "Fog"

userInput = [[1,7,2.5,2.3]]
print("Today's weather is " + predictWeather(userInput))

Today's weather is Rain


### **Summary**

*   1st model accuracy : 73.04 %
*   2nd model accuracy : 75.09 %
*   3rd model accuracy : 78.84 %
*   4th model accuracy : 83.28 %
*   5th model accuracy : 83.28 %
*   6th model accuracy : 78.84 %
*   7th model accuracy : 83.28 %
*   8th model accuracy : 83.28 %


### **Observation**

Train and test the different model with same dataset we can observe that we need to remove outliers and also scale our data to improve time complexity. We also need to set some parameters like maximum depth of decision tree and minimum sample leaf node so that get good results.